src/file_selector.ml

```diff
@@ -0,0 +1,39 @@
+open Stdune
+
+type t =
+  { dir : Path.t
+  ; predicate : Path.t Predicate.t
+  }
+
+let dir t = t.dir
+
+let compare x y =
+  match Path.compare x.dir y.dir with
+  | Ordering.Lt | Gt as a -> a
+  | Eq -> Predicate.compare x.predicate y.predicate
+
+let create ~dir predicate = { dir ; predicate }
+
+let to_dyn { dir ; predicate } =
+  let open Dyn in
+  Record
+    [ "dir", Path.to_dyn dir
+    ; "predicate", Predicate.to_dyn predicate
+    ]
+
+let pp fmt t = Dyn.pp fmt (to_dyn t)
+
+let encode { dir; predicate } =
+  let open Dune_lang.Encoder in
+  record
+    [ "dir", Path_dune_lang.encode dir
+    ; "predicate", Predicate.encode predicate
+    ]
+
+let equal x y = compare x y = Eq
+let hash { dir; predicate} =
+  Tuple.T2.hash Path.hash Predicate.hash (dir, predicate)
+
+let to_sexp t = Dyn.to_sexp (to_dyn t)
+
+let test t = Predicate.test t.predicate
```

Comment:
I think it might be worth checking here that the path is a child of `t.dir` and raise an exception if it is not. (assuming we agree that this type is only meaningful within a single directory)